File: word4you-python/gemini_client.py
from google import genai
from google.genai import types
from config import Config
import time

class GeminiClient:
    """Client for interacting with Google Gemini API."""
    
    def __init__(self):
        """Initialize the Gemini client."""
        self.client = genai.Client(api_key=Config.GEMINI_API_KEY)
    
    def get_word_explanation(self, word: str) -> str:
        """
        Get word explanation from Gemini API.
        
        Args:
            word: The English word to explain
            
        Returns:
            str: Markdown formatted explanation
        """
        try:
            # Format the prompt with the word
            prompt = Config.GEMINI_PROMPT_TEMPLATE.format(word=word.lower())
            
            # Generate response using the new SDK
            response = self.client.models.generate_content(
                model='gemini-2.0-flash-001',
                contents=prompt
            )
            
            if response.text:
                return response.text
            else:
                raise Exception("No response received from Gemini API")
                
        except Exception as e:
            raise Exception(f"Error getting word explanation: {str(e)}")
    
    def test_connection(self) -> bool:
        """Test the connection to Gemini API."""
        try:
            response = self.client.models.generate_content(
                model='gemini-2.0-flash-001',
                contents="Hello"
            )
            return response.text is not None
        except Exception:
            return False 

File: word4you-cli/src/gemini_client.rs
use anyhow::{anyhow, Result};
use reqwest::Client;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize)]
struct GeminiRequest {
    contents: Vec<Content>,
}

#[derive(Debug, Serialize)]
struct Content {
    parts: Vec<Part>,
}

#[derive(Debug, Serialize)]
struct Part {
    text: String,
}

#[derive(Debug, Deserialize)]
struct GeminiResponse {
    candidates: Vec<Candidate>,
}

#[derive(Debug, Deserialize)]
struct Candidate {
    content: ContentResponse,
}

#[derive(Debug, Deserialize)]
struct ContentResponse {
    parts: Vec<PartResponse>,
}

#[derive(Debug, Deserialize)]
struct PartResponse {
    text: String,
}

pub struct GeminiClient {
    client: Client,
    api_key: String,
    base_url: String,
}

impl GeminiClient {
    pub fn new(api_key: String, model_name: String) -> Self {
        let base_url = format!(
            "https://generativelanguage.googleapis.com/v1beta/models/{}:generateContent",
            model_name
        );
        Self {
            client: Client::new(),
            api_key,
            base_url,
        }
    }

    pub async fn get_word_explanation(&self, word: &str, prompt_template: &str) -> Result<String> {
        let prompt = prompt_template.replace("[INSERT WORD HERE]", &word.to_lowercase());

        let request = GeminiRequest {
            contents: vec![Content {
                parts: vec![Part { text: prompt }],
            }],
        };

        let url = format!("{}?key={}", self.base_url, self.api_key);

        let response = self.client.post(&url).json(&request).send().await?;

        if !response.status().is_success() {
            let error_text = response.text().await?;
            return Err(anyhow!("Gemini API error: {}", error_text));
        }

        let gemini_response: GeminiResponse = response.json().await?;

        if let Some(candidate) = gemini_response.candidates.first() {
            if let Some(part) = candidate.content.parts.first() {
                return Ok(part.text.clone().trim().to_string());
            }
        }

        Err(anyhow!("No response received from Gemini API"))
    }

    pub async fn test_connection(&self) -> Result<bool> {
        let request = GeminiRequest {
            contents: vec![Content {
                parts: vec![Part {
                    text: "Hello".to_string(),
                }],
            }],
        };

        let url = format!("{}?key={}", self.base_url, self.api_key);

        let response = self.client.post(&url).json(&request).send().await;

        match response {
            Ok(resp) => {
                if resp.status().is_success() {
                    Ok(true)
                } else {
                    Ok(false)
                }
            }
            Err(_) => Ok(false),
        }
    }
}
